odbc/src/api.rs

```diff
@@ -7,11 +11,49 @@ use odbc_sys::{
 
 #[no_mangle]
 pub extern "C" fn SQLAllocHandle(
-    _handle_type: HandleType,
-    _input_handle: Handle,
-    _output_handle: *mut Handle,
-) -> SqlReturn {
-    unimplemented!()
+    handle_type: HandleType,
+    input_handle: Handle,
+    output_handle: *mut Handle,
+) -> SqlReturn {
+    match handle_type {
+        HandleType::Env => {
+            let env = Box::new(EnvHandle::new());
+            env.env.write().unwrap().state = EnvState::Allocated;
+            unsafe { *output_handle = Box::into_raw(env) as *mut _ }
+        }
+        HandleType::Dbc => {
+            let env = input_handle as *mut _;
+            let conn = Box::new(ConnectionHandle::new(env));
```

Comment:
We may, here, wish to return an error if we try to allocate a connection on an unallocated environment, but we can leave that for when we have error support.

Reply:
sgtm